File: frontend/mdfe/lib/app/data/model/mdfe_rodoviario_pedagio_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:mdfe/app/infra/infra_imports.dart';
import 'package:mdfe/app/data/model/model_imports.dart';

class MdfeRodoviarioPedagioModel {
	int? id;
	int? idMdfeRodoviario;
	String? cnpjFornecedor;
	String? cnpjResponsavel;
	String? cpfResponsavel;
	String? numeroComprovante;
	double? valor;
	MdfeRodoviarioModel? mdfeRodoviarioModel;

	MdfeRodoviarioPedagioModel({
		this.id,
		this.idMdfeRodoviario,
		this.cnpjFornecedor,
		this.cnpjResponsavel,
		this.cpfResponsavel,
		this.numeroComprovante,
		this.valor,
		this.mdfeRodoviarioModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'cnpj_fornecedor',
		'cnpj_responsavel',
		'cpf_responsavel',
		'numero_comprovante',
		'valor',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Cnpj Fornecedor',
		'Cnpj Responsavel',
		'Cpf Responsavel',
		'Numero Comprovante',
		'Valor',
	];

	MdfeRodoviarioPedagioModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idMdfeRodoviario = jsonData['idMdfeRodoviario'];
		cnpjFornecedor = jsonData['cnpjFornecedor'];
		cnpjResponsavel = jsonData['cnpjResponsavel'];
		cpfResponsavel = jsonData['cpfResponsavel'];
		numeroComprovante = jsonData['numeroComprovante'];
		valor = jsonData['valor']?.toDouble();
		mdfeRodoviarioModel = jsonData['mdfeRodoviarioModel'] == null ? MdfeRodoviarioModel() : MdfeRodoviarioModel.fromJson(jsonData['mdfeRodoviarioModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idMdfeRodoviario'] = idMdfeRodoviario != 0 ? idMdfeRodoviario : null;
		jsonData['cnpjFornecedor'] = Util.removeMask(cnpjFornecedor);
		jsonData['cnpjResponsavel'] = Util.removeMask(cnpjResponsavel);
		jsonData['cpfResponsavel'] = Util.removeMask(cpfResponsavel);
		jsonData['numeroComprovante'] = numeroComprovante;
		jsonData['valor'] = valor;
		jsonData['mdfeRodoviarioModel'] = mdfeRodoviarioModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idMdfeRodoviario = plutoRow.cells['idMdfeRodoviario']?.value;
		cnpjFornecedor = plutoRow.cells['cnpjFornecedor']?.value;
		cnpjResponsavel = plutoRow.cells['cnpjResponsavel']?.value;
		cpfResponsavel = plutoRow.cells['cpfResponsavel']?.value;
		numeroComprovante = plutoRow.cells['numeroComprovante']?.value;
		valor = plutoRow.cells['valor']?.value?.toDouble();
		mdfeRodoviarioModel = MdfeRodoviarioModel();
		mdfeRodoviarioModel?.codigoAgendamento = plutoRow.cells['mdfeRodoviarioModel']?.value;
	}	

	MdfeRodoviarioPedagioModel clone() {
		return MdfeRodoviarioPedagioModel(
			id: id,
			idMdfeRodoviario: idMdfeRodoviario,
			cnpjFornecedor: cnpjFornecedor,
			cnpjResponsavel: cnpjResponsavel,
			cpfResponsavel: cpfResponsavel,
			numeroComprovante: numeroComprovante,
			valor: valor,
		);			
	}

	
}

File: backend/mdfe/src/model/mdfe_rodoviario_pedagio_model.py
from src import db
from src.model.mdfe_rodoviario_model import MdfeRodoviarioModel


class MdfeRodoviarioPedagioModel(db.Model):
    __tablename__ = 'mdfe_rodoviario_pedagio'

    id = db.Column(db.Integer, primary_key=True)
    cnpj_fornecedor = db.Column(db.String(14))
    cnpj_responsavel = db.Column(db.String(14))
    cpf_responsavel = db.Column(db.String(11))
    numero_comprovante = db.Column(db.String(20))
    valor = db.Column(db.Float)
    id_mdfe_rodoviario = db.Column(db.Integer, db.ForeignKey('mdfe_rodoviario.id'))

    mdfe_rodoviario_model = db.relationship('MdfeRodoviarioModel', foreign_keys=[id_mdfe_rodoviario])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_mdfe_rodoviario = data.get('idMdfeRodoviario')
        self.cnpj_fornecedor = data.get('cnpjFornecedor')
        self.cnpj_responsavel = data.get('cnpjResponsavel')
        self.cpf_responsavel = data.get('cpfResponsavel')
        self.numero_comprovante = data.get('numeroComprovante')
        self.valor = data.get('valor')

    def serialize(self):
        return {
            'id': self.id,
            'idMdfeRodoviario': self.id_mdfe_rodoviario,
            'cnpjFornecedor': self.cnpj_fornecedor,
            'cnpjResponsavel': self.cnpj_responsavel,
            'cpfResponsavel': self.cpf_responsavel,
            'numeroComprovante': self.numero_comprovante,
            'valor': self.valor,
            'mdfeRodoviarioModel': self.mdfe_rodoviario_model.serialize() if self.mdfe_rodoviario_model else None,
        }